src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -324,13 +331,40 @@ protected override void Dispose(bool disposing)
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
             }
 
             this.textWriter.Dispose();
             base.Dispose(disposing);
         }
+
+#if NETSTANDARD2_0
+        public async ValueTask DisposeAsync()
+        {
+            if (this.listener != null)
+            {
+                await this.listener.StreamDisposedAsync()
+                    .ConfigureAwait(false);
+
+                this.listener = null;
+            }
+
+            // Dispose unmanaged resources
+            // Pass false to ensure functional equivalence with the synchronous dispose pattern
+            this.Dispose(false);
+            // Suppress finalization
+            GC.SuppressFinalize(this);

```

Comment:
Do we need to suppress finalization if this class doesn't implement a finalizer? But I guess it doesn't hurt to call either way.

Reply:
I followed the guideline [here](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-the-async-dispose-pattern). Note however that in our class we don't need to implement `DisposeAsyncCore` since the class is sealed.